### PR TITLE
generate metadata from throwable

### LINKED ISSUE #4203: grpc-status-details-bin is not propagated in response headers when throw exception specifically
In grpc AbstractUnframedGrpcService, when we throw errors other than doing `responseObserver.onError(StatusProto.toStatusRuntimeException(status))`, the `grpc-status-details-bin` will not be propagated in headers. 
eg. when we do something like 
```
final ErrorInfo errorInfo = ErrorInfo.newBuilder()
                                     .setDomain("test")
                                     .setReason("Unknown Exception").build();

final com.google.rpc.Status
    status = com.google.rpc.Status.newBuilder()
                                  .setCode(
                                      Code.UNKNOWN.getNumber())
                                  .setMessage("Unknown Exceptions Test")
                                  .addDetails(
                                      Any.pack(errorInfo))
                                  .build();

throw StatusProto.toStatusRuntimeException(status);
```
the metadata information is lost in https://github.com/line/armeria/blob/f0a7a01f79e54cef847699091cbe7d7ba94ae3a2/grpc/src/main/java/com/linecorp/armeria/server/grpc/ArmeriaServerCall.java#L567

### PR DESCRIPTION
**Pull request by natsumehu**
Motivation:

Meta data information is missing when throw exception specifically. We wanna keep the grpc-status-details-bin whenever possible.

Modifications:

- add a generateMetadataFromThrowable method to be called when exception happens

Result:

- Closes #4203.

### CODE REVIEW

**Review by trustin — COMMENTED** [review]
Thanks a lot, @natsumehu! Looks like a correct fix. Would you mind adding a test case?

File: grpc/src/main/java/com/linecorp/armeria/server/grpc/ArmeriaServerCall.java [hunk 1]
```diff
@@ -564,7 +564,7 @@ private void invokeHalfClose() {
             assert listener != null;
             listener.onHalfClose();
         } catch (Throwable t) {
-            close(t, new Metadata());
+            close(t, generateMetadataFromThrowable(t));
```
Comment (ikhoon) [inline grpc/src/main/java/com/linecorp/armeria/server/grpc/ArmeriaServerCall.java:567]: How about introducing `close(Throwable)` method and replace all `close(t, new Metadata())` with the new one?

**Comment by natsumehu** [comment]
I think I might need to wait https://github.com/line/armeria/pull/4171 to be merged

**Comment by ikhoon** [comment]
It will be merged soon. 😉

**Comment by trustin** [comment]
@natsumehu It's been merged. Thanks for waiting. Would you mind resolving a merge conflict?

**Comment by natsumehu** [comment]
> @natsumehu It's been merged. Thanks for waiting. Would you mind resolving a merge conflict?

Sure, will work on it asap

**Review by trustin — APPROVED** [review]
Nice work! :+1:

File: grpc/src/main/java/com/linecorp/armeria/server/grpc/UnframedGrpcErrorHandler.java [hunk 2]
```diff
@@ -102,11 +105,17 @@ static UnframedGrpcErrorHandler ofJson(UnframedGrpcStatusMappingFunction statusM
...
+                    ResponseHeaders.builder(httpStatus)
+                                   .contentType(MediaType.JSON_UTF_8)
+                                   .addInt(GrpcHeaderNames.GRPC_STATUS,
+                                           grpcCode.value());
+            final HttpHeaders trailers = !response.trailers().isEmpty() ?
+                                         response.trailers() : response.headers();
+            final String grpcStatusDetailsBin = trailers.get(GrpcHeaderNames.GRPC_STATUS_DETAILS_BIN);
+            if (!Strings.isNullOrEmpty(grpcStatusDetailsBin)) {
+                responseHeadersBuilder.set(GrpcHeaderNames.GRPC_STATUS_DETAILS_BIN, grpcStatusDetailsBin);
+            }
```
Comment (ikhoon) [inline grpc/src/main/java/com/linecorp/armeria/server/grpc/UnframedGrpcErrorHandler.java:118]: Shouldn't we need to render a rich error response using the `GrpcHeaderNames.GRPC_STATUS_DETAILS_BIN`?
https://cloud.google.com/apis/design/errors#error_model
I guess users don't want to deserialize Protobuf data in headers.
https://cloud.google.com/apis/design/errors#error_model
```json
{
  "error": {
    "code": 400,
    "message": "API key not valid. Please pass a valid API key.",
    "status": "INVALID_ARGUMENT",
    "details": [
      {
        "@type": "type.googleapis.com/google.rpc.ErrorInfo",
        "reason": "API_KEY_INVALID",
        "domain": "googleapis.com",
        "metadata": {
          "service": "translate.googleapis.com"
        }
      }
    ]
  }
}
```

Comment (ikhoon) [reply grpc/src/main/java/com/linecorp/armeria/server/grpc/UnframedGrpcErrorHandler.java:118]: On top of that, is it better to modify the default error response format for better compatibility with google APIs?
For example:
* if `GRPC_STATUS_DETAILS_BIN is not specified
```json
{
  "error": {
    "code": 400,
    "message": "API key not valid. Please pass a valid API key.",
    "status": "INVALID_ARGUMENT"
  }
}
```
* if `GRPC_STATUS_DETAILS_BIN is specified
```json
{
  "error": {
    "code": 400,
    "message": "API key not valid. Please pass a valid API key.",
    "status": "INVALID_ARGUMENT",
    "details": [ ... ]
  }
}
```

Comment (natsumehu) [reply grpc/src/main/java/com/linecorp/armeria/server/grpc/UnframedGrpcErrorHandler.java:118]: Oh, sure. I originally planned to include the rich error model in my next PR. But yes, I would finish that later on in this PR. Thanks for the advise.

Comment (ikhoon) [reply grpc/src/main/java/com/linecorp/armeria/server/grpc/UnframedGrpcErrorHandler.java:118]: Do you think is it useful to include `grpc-status-detail-bin` in response headers? I thought we don't need to propagate the raw binary header.

Comment (ikhoon) [reply grpc/src/main/java/com/linecorp/armeria/server/grpc/UnframedGrpcErrorHandler.java:118]: If you are going to handle the rich error model, how about reverting the changes in `UnframedGrpcErrorHandler` and handling it together in the next PR?

Comment (natsumehu) [reply grpc/src/main/java/com/linecorp/armeria/server/grpc/UnframedGrpcErrorHandler.java:118]: > If you are going to handle the rich error model, how about reverting the changes in `UnframedGrpcErrorHandler` and handling it together in the next PR?

Oh, sure. I will think about whether to include grpc-status-detail-bin in response headers later on. And can you advise me on how to write a test case if I do not include that in response headers for this PR?

Comment (ikhoon) [reply grpc/src/main/java/com/linecorp/armeria/server/grpc/UnframedGrpcErrorHandler.java:118]: I think we can capture a `ResponseHeaders` by adding a server-slide decorator.

Comment (natsumehu) [reply grpc/src/main/java/com/linecorp/armeria/server/grpc/UnframedGrpcErrorHandler.java]: @ikhoon  reverted.

**Review by ikhoon — APPROVED** [review]
Thanks, @natsumehu!

**Pull request merged**